inference-engine/src/vpu/graph_transformer/src/middleend/special_stage_processor.cpp

```diff
@@ -5,99 +5,249 @@
         }
 
         //
-        // Insert Copy if needed
+        // Check producer StridesRequirement.
         //
 
-        if (needCopy) {
-            auto outputCopy = model->duplicateData(output, "@copy");
-            outputCopy->resetRequiredStrides();
+        if (!isCopyNeed) {
+            if (auto producerEdge = input->producerEdge()) {
+                const auto& producerInfo = producerEdge->producer()->getDataStridesRequirements();
 
-            auto outPortInd = outEdge->portInd();
+                if (producerInfo.hasOutput(producerEdge)) {
+                    const auto& producerStrideReqs = producerInfo.getOutput(producerEdge);
+                    IE_ASSERT(input->checkStrides(producerStrideReqs));
 
-            model->replaceStageOutput(outEdge, outputCopy);
+                    if (!checkStrides(input->desc(), outputData->strides(), producerStrideReqs)) {
+                        isCopyNeed = true;
+                        isCopyOptional = false;
+                    }
+                }
 
-            auto copyStage = _stageBuilder->addCopyStage(
-                model,
-                formatString("%s@output=%d@copy-for-split", stage->name(), outPortInd),
-                stage->origLayer(),
-                outputCopy,
-                output,
-                "special::split");
-            if (stage->attrs().has("batchInd")) {
-                copyStage->attrs().set("batchInd", stage->attrs().get<int>("batchInd"));
+                if (!isCopyNeed) {
+                    //
+                    // To reduce the size of HW output (still can be optimized).
+                    //
+
+                    if (producerEdge->producer()->category() == StageCategory::HW) {
+                        isCopyNeed = true;
+                        isCopyOptional = true;
+                    }
+                }
             }
+        }
+    }
 
-            output = outputCopy;
+    NeedCopyDesc needCopyDesc;
+    needCopyDesc.isCopyNeed = isCopyNeed;
+    if (stage->type() == StageType::StubConcat ||
+        stage->type() == StageType::Expand) {
+        needCopyDesc.isCopyOptional = isCopyOptional;
+    }
+    return needCopyDesc;
+}
+
+Data insertCopyOfInput(
+                const Model& model,
+                const Stage& stage,
+                const StageInput& edge,
+                const StageBuilder::Ptr& _stageBuilder,
+                const char* stageType,
+                const NeedCopyDesc& desc) {
+    auto data = edge->input();
+    Data copy;
+    if (data->usage() == DataUsage::Const) {
+        copy = model->addNewData(
+                data->name() + "@copy",
+                data->desc());
+    } else {
+        copy = model->duplicateData(
+                data,
+                "@copy");
+        copy->resetRequiredStrides();
+    }
+    if (stage->type() == StageType::Reshape)
+        copy->updateRequiredStrides(StridesRequirement::compact());
+
+    bool isManyInputs = stage->numInputs() > 1;
```

Comment:
Nitpick: `isManyInputs` -> `hasMultipleInputs`

Reply:
Actually I think we can do the same for both cases

Reply:
I guess than we have no need to make ambiguity and overload .dot graph representation.

Reply:
replaced `isManyInputs` to `hasMultipleInputs`